File: android/app/src/main/kotlin/jp/co/bitz/tictactoe/game/states/TurnEndState.kt
package jp.co.bitz.tictactoe.game.states

import jp.co.bitz.gameplaykit.GKState
import jp.co.bitz.tictactoe.game.TicTacToeMatch
import kotlin.reflect.KClass

/**
 * Checks whether the game just ended; routes to [GameOverState] or back to [TurnBeginState] for
 * the next turn. See docs/GAME_DESIGN.md's "Turn flow" section.
 */
public class TurnEndState(
    private val match: TicTacToeMatch,
) : GKState() {
    override fun didEnter(previousState: GKState?) {
        if (match.gameModel.board.isGameOver) {
            stateMachine?.enter<GameOverState>()
        } else {
            stateMachine?.enter<TurnBeginState>()
        }
    }

    override fun isValidNextState(stateClass: KClass<out GKState>): Boolean =
        stateClass == GameOverState::class || stateClass == TurnBeginState::class
}

File: ios/TicTacToe/Game/States/TurnEndState.swift
import GameplayKit

/// Checks whether the game just ended; routes to `GameOverState` or back to `TurnBeginState` for
/// the next turn. See `docs/GAME_DESIGN.md`'s "Turn flow" section.
final class TurnEndState: GKState {
    unowned let match: TicTacToeMatch

    init(match: TicTacToeMatch) {
        self.match = match
    }

    override func didEnter(from previousState: GKState?) {
        if match.gameModel.board.isGameOver {
            _ = stateMachine?.enter(GameOverState.self)
        } else {
            _ = stateMachine?.enter(TurnBeginState.self)
        }
    }

    override func isValidNextState(_ stateClass: AnyClass) -> Bool {
        stateClass == GameOverState.self || stateClass == TurnBeginState.self
    }
}
